### ISSUE TITLE
screener

### ISSUE THREAD

**Issue by schcerbakov** [schcerbakov/pc#1]
import requests
import time
import asyncio
from telegram import Update
from telegram.ext import ApplicationBuilder, CommandHandler, ContextTypes

TOKEN = "8332409278:AAHRui4MzkoEGqr_iRSWpL_NsRMIOVzgZbg"
ALERT_CHAT_ID = "1415500433"

# =========================
# 📊 STATE
# =========================
previous_snapshot = {}
last_update = 0

# =========================
# 📡 BYBIT API
# =========================
def fetch_bybit_data():
    url = "https://api.bybit.com/v5/market/tickers?category=linear"
    r = requests.get(url, timeout=15)
    data = r.json()

    if data.get("retCode") != 0:
        raise Exception(data)

    return data.get("result", {}).get("list", [])

# =========================
# 💾 SNAPSHOT
# =========================
def update_snapshot(coins):
    global previous_snapshot, last_update

    previous_snapshot = {
        c.get("symbol"): float(c.get("turnover24h") or 0)
        for c in coins
        if c.get("symbol")
    }

    last_update = time.time()

# =========================
# 🧠 FILTER (ANTI NOISE)
# =========================
def is_valid(delta_m, price_change):
    return delta_m >= 3 and abs(price_change) >= 1.5

# =========================
# 🚀 START
# =========================
async def start(update: Update, context: ContextTypes.DEFAULT_TYPE):
    await update.message.reply_text(
        "🚀 Bybit Scanner Bot\n\n"
        "/top - рынок\n"
        "/change - low cap volume spikes\n"
    )

# =========================
# 📊 TOP MARKET
# =========================
async def top(update: Update, context: ContextTypes.DEFAULT_TYPE):
    try:
        coins = fetch_bybit_data()

        volume = []
        movers = []

        for c in coins:
            sym = c.get("symbol", "")

            if sym.endswith("USDT"):
                vol = float(c.get("turnover24h") or 0)
                chg = float(c.get("price24hPcnt") or 0) * 100

                volume.append({"symbol": sym, "vol": vol / 1_000_000})
                movers.append({"symbol": sym, "chg": chg})

        volume = sorted(volume, key=lambda x: x["vol"], reverse=True)[:15]
        gain = sorted(movers, key=lambda x: x["chg"], reverse=True)[:10]
        loss = sorted(movers, key=lambda x: x["chg"])[:10]

        msg = "📊 TOP MARKET\n\n"

        msg += "📦 Volume:\n"
        for x in volume:
            msg += f"{x['symbol']} | {x['vol']:.2f}M\n"

        msg += "\n📈 Gainers:\n"
        for x in gain:
            msg += f"{x['symbol']} | +{x['chg']:.2f}%\n"

        msg += "\n📉 Losers:\n"
        for x in loss:
            msg += f"{x['symbol']} | {x['chg']:.2f}%\n"

        await update.message.reply_text(msg)

    except Exception as e:
        await update.message.reply_text(str(e))

# =========================
# 🚨 /CHANGE (LOW CAP ENGINE)
# =========================
async def change(update: Update, context: ContextTypes.DEFAULT_TYPE):
    global previous_snapshot, last_update

    try:
        coins = fetch_bybit_data()

        # init snapshot
        if not previous_snapshot:
            update_snapshot(coins)
            await update.message.reply_text("⏳ Инициализация базы... повтори через 1 минуту")
            return

        results = []

        for c in coins:
            symbol = c.get("symbol", "")

            if not symbol.endswith("USDT"):
                continue

            current_volume = float(c.get("turnover24h") or 0)

            # 📌 LOW CAP FILTER (<10M)
            if current_volume >= 10_000_000:
                continue

            previous_volume = previous_snapshot.get(symbol, 0)

            delta_m = (current_volume - previous_volume) / 1_000_000
            price_change = float(c.get("price24hPcnt") or 0) * 100

            # 🚨 SIGNAL FILTER
            if is_valid(delta_m, price_change):
                results.append({
                    "symbol": symbol,
                    "delta": delta_m,
                    "volume": current_volume / 1_000_000,
                    "price": price_change
                })

        results = sorted(results, key=lambda x: x["delta"], reverse=True)

        message = "🚨 LOW CAP VOLUME SPIKE (<10M)\n\n"

        if not results:
            message += "Нет сильных движений сейчас."
        else:
            for r in results[:15]:
                message += (
                    f"{r['symbol']} | "
                    f"+{r['delta']:.2f}M Δ | "
                    f"Vol: {r['volume']:.2f}M | "
                    f"{r['price']:+.2f}%\n"
                )

        await update.message.reply_text(message)

        update_snapshot(coins)

    except Exception as e:
        await update.message.reply_text(f"Ошибка: {e}")

# =========================
# 🤖 BOT START
# =========================
app = ApplicationBuilder().token(TOKEN).build()

app.add_handler(CommandHandler("start", start))
app.add_handler(CommandHandler("top", top))
app.add_handler(CommandHandler("change", change))

print("Bot running...")
app.run_polling()
